File: src/Library/ObjectModel/RegionSetupStep.vb
Public Class RegionSetupStep
    Inherits Pixa.Soundbridge.Library.SetupStep

    Friend Sub New(ByVal client As ISoundbridgeClient)
        MyBase.New(client)
    End Sub

    Protected Overrides ReadOnly Property Name() As String
        Get
            Return "Region"
        End Get
    End Property

    Public Overrides Function GetSelectionList() As String()
        Return Client.ListRegions()
    End Function

    Public Overrides Sub MakeSelection(ByVal value As String)
        Dim regions As String() = Client.ListRegions
        Dim index As Integer = Array.IndexOf(regions, value)

        If index = -1 Then Throw New ArgumentException("Specified Region was not in the list of valid regions")

        Dim r As String = Client.SetRegion(index)

        If r = "OK" Then Exit Sub
        If r = "ParameterError" Then Throw New ArgumentException("Invalid index or list results", value)
        If r = "ErrorUnsupported" Then Throw New ArgumentException("The specified region is not supported on this host", value)
        If r = "ErrorAlreadySet" Then Throw New ArgumentException("The region has already been set", value)
        If r = "GenericError" Then Throw New ArgumentException("Unable to set region", value)
    End Sub
End Class

File: src/Pixa.Soundbridge/RegionSetupStep.cs
using Pixa.Soundbridge.Client;
using System;

namespace Pixa.Soundbridge {
    public class RegionSetupStep : SetupStep {
        internal RegionSetupStep(ISoundbridgeClient client) : base(client) {
        }

        protected override string Name {
            get {
                return "Region";
            }
        }

        public override string[] GetSelectionList() {
            return Client.ListRegions();
        }

        public override void MakeSelection(string value) {
            var regions = Client.ListRegions();
            int index = Array.IndexOf(regions, value);
            if (index == -1)
                throw new ArgumentException("Specified Region was not in the list of valid regions");
            string r = Client.SetRegion(index);
            if (r == "OK")
                return;
            if (r == "ParameterError")
                throw new ArgumentException("Invalid index or list results", value);
            if (r == "ErrorUnsupported")
                throw new ArgumentException("The specified region is not supported on this host", value);
            if (r == "ErrorAlreadySet")
                throw new ArgumentException("The region has already been set", value);
            if (r == "GenericError")
                throw new ArgumentException("Unable to set region", value);
        }
    }
}